File: langs/ruby/lib/dotlyte/validator.rb
# frozen_string_literal: true

module Dotlyte
  # Schema rule for a single config key.
  class SchemaRule
    attr_accessor :type, :required, :format, :pattern, :enum_values,
                  :min, :max, :default_value, :sensitive, :doc

    def initialize(**kwargs)
      @type = kwargs[:type]
      @required = kwargs.fetch(:required, false)
      @format = kwargs[:format]
      @pattern = kwargs[:pattern]
      @enum_values = kwargs[:enum_values]
      @min = kwargs[:min]
      @max = kwargs[:max]
      @default_value = kwargs[:default_value]
      @sensitive = kwargs.fetch(:sensitive, false)
      @doc = kwargs[:doc]
    end
  end

  # A single schema violation.
  class SchemaViolation
    attr_reader :key, :message, :rule

    def initialize(key:, message:, rule:)
      @key = key
      @message = message
      @rule = rule
    end

    def to_s
      "[#{rule}] #{key}: #{message}"
    end
  end

  # Schema validation engine for DOTLYTE v2.
  module Validator
    FORMAT_PATTERNS = {
      "email" => /\A[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}\z/,
      "uuid"  => /\A[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\z/,
      "date"  => /\A\d{4}-\d{2}-\d{2}\z/,
      "ipv4"  => /\A\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\z/
    }.freeze

    # Validate config data against a schema.
    #
    # @param data [Hash] the config data
    # @param schema [Hash<String, SchemaRule>] the schema
    # @param strict [Boolean] reject unknown keys
    # @return [Array<SchemaViolation>]
    def self.validate(data, schema, strict: false)
      violations = []

      schema.each do |key, rule|
        val = get_nested(data, key)

        if val.nil?
          if rule.required
            violations << SchemaViolation.new(
              key: key, message: "required key '#{key}' is missing", rule: "required"
            )
          end
          next
        end

        # Type check
        if rule.type && !check_type(val, rule.type)
          violations << SchemaViolation.new(
            key: key,
            message: "expected type '#{rule.type}', got #{val.class}",
            rule: "type"
          )
        end

        # Format check
        if rule.format && val.is_a?(String) && !check_format(val, rule.format)
          violations << SchemaViolation.new(
            key: key,
            message: "value '#{val}' does not match format '#{rule.format}'",
            rule: "format"
          )
        end

        # Pattern check
        if rule.pattern && val.is_a?(String) && !val.match?(Regexp.new(rule.pattern))
          violations << SchemaViolation.new(
            key: key,
            message: "value '#{val}' does not match pattern '#{rule.pattern}'",
            rule: "pattern"
          )
        end

        # Enum check
        if rule.enum_values && !rule.enum_values.include?(val)
          violations << SchemaViolation.new(
            key: key,
            message: "value #{val} not in allowed values: #{rule.enum_values}",
            rule: "enum"
          )
        end

        # Min/Max
        if val.is_a?(Numeric)
          if rule.min && val < rule.min
            violations << SchemaViolation.new(
              key: key, message: "value #{val} is less than minimum #{rule.min}", rule: "min"
            )
          end
          if rule.max && val > rule.max
            violations << SchemaViolation.new(
              key: key, message: "value #{val} is greater than maximum #{rule.max}", rule: "max"
            )
          end
        end
      end

      # Strict mode
      if strict
        flat_keys = flatten_keys(data)
        flat_keys.each do |k|
          unless schema.key?(k)
            violations << SchemaViolation.new(
              key: k, message: "unknown key '#{k}' (strict mode)", rule: "strict"
            )
          end
        end
      end

      violations
    end

    # Apply schema defaults to data.
    def self.apply_defaults(data, schema)
      schema.each do |key, rule|
        next if rule.default_value.nil?
        next unless get_nested(data, key).nil?

        set_nested(data, key, rule.default_value)
      end
    end

    # Get all sensitive keys from schema.
    def self.sensitive_keys(schema)
      schema.select { |_, rule| rule.sensitive }.keys
    end

    # Assert valid — raises ValidationError on failure.
    def self.assert_valid!(data, schema, strict: false)
      violations = validate(data, schema, strict: strict)
      return if violations.empty?

      msg = "Schema validation failed:\n" +
            violations.map { |v| "  - #{v}" }.join("\n")
      raise ValidationError.new(msg, violations: violations)
    end

    # Get a nested value via dot-notation key.
    def self.get_nested(data, key)
      parts = key.to_s.split(".")
      current = data
      parts.each do |part|
        return nil unless current.is_a?(Hash)

        current = current.key?(part) ? current[part] : current[part.to_sym]
        return nil if current.nil?
      end
      current
    end

    # Set a nested value via dot-notation key.
    def self.set_nested(data, key, value)
      parts = key.to_s.split(".")
      current = data
      parts[0..-2].each do |part|
        current[part] ||= {}
        current[part] = {} unless current[part].is_a?(Hash)
        current = current[part]
      end
      current[parts.last] = value
    end

    class << self
      private

      def check_type(val, expected)
        case expected
        when "string"  then val.is_a?(String)
        when "number"  then val.is_a?(Numeric)
        when "boolean" then [true, false].include?(val)
        when "array"   then val.is_a?(Array)
        when "hash", "object" then val.is_a?(Hash)
        else true
        end
      end

      def check_format(val, fmt)
        case fmt
        when "url"
          val.start_with?("http://") || val.start_with?("https://")
        when "ip", "ipv4"
          FORMAT_PATTERNS["ipv4"].match?(val)
        when "port"
          p = Integer(val, exception: false)
          p && p >= 1 && p <= 65_535
        when "email"
          FORMAT_PATTERNS["email"].match?(val)
        when "uuid"
          FORMAT_PATTERNS["uuid"].match?(val)
        when "date"
          FORMAT_PATTERNS["date"].match?(val)
        else
          val.match?(Regexp.new(fmt))
        end
      end

      def flatten_keys(data, prefix = "", out = [])
        data.each do |key, value|
          full_key = prefix.empty? ? key.to_s : "#{prefix}.#{key}"
          if value.is_a?(Hash)
            flatten_keys(value, full_key, out)
          else
            out << full_key
          end
        end
        out
      end
    end
  end
end

File: langs/dotnet/src/Dotlyte/Validator.cs
using System.Text.RegularExpressions;

namespace Dotlyte;

/// <summary>
/// Schema validation engine for DOTLYTE v2.
/// </summary>
public static class Validator
{
    private static readonly Dictionary<string, string> FormatPatterns = new()
    {
        ["email"] = @"^[^@\s]+@[^@\s]+\.[^@\s]+$",
        ["uuid"] = @"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$",
        ["date"] = @"^\d{4}-\d{2}-\d{2}$",
        ["ipv4"] = @"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$",
    };

    /// <summary>
    /// Validate data against a schema. Returns a list of violations.
    /// </summary>
    public static List<SchemaViolation> Validate(
        Dictionary<string, object?> data,
        Dictionary<string, SchemaRule> schema,
        bool strict = false)
    {
        var violations = new List<SchemaViolation>();

        foreach (var (key, rule) in schema)
        {
            var value = GetNested(data, key);

            if (value is null)
            {
                if (rule.Required)
                    violations.Add(new SchemaViolation(key, $"Required key '{key}' is missing", rule));
                continue;
            }

            if (rule.Type is not null && !CheckType(value, rule.Type))
                violations.Add(new SchemaViolation(key,
                    $"Expected type '{rule.Type}' but got '{value.GetType().Name}'", rule));

            if (rule.Format is not null && value is string strVal && !CheckFormat(strVal, rule.Format))
                violations.Add(new SchemaViolation(key,
                    $"Value does not match format '{rule.Format}'", rule));

            if (rule.Pattern is not null && value is string patVal && !Regex.IsMatch(patVal, rule.Pattern))
                violations.Add(new SchemaViolation(key,
                    $"Value does not match pattern '{rule.Pattern}'", rule));

            if (rule.EnumValues is not null)
            {
                var strValue = value.ToString();
                if (!rule.EnumValues.Contains(strValue))
                    violations.Add(new SchemaViolation(key,
                        $"Value '{strValue}' not in allowed values: [{string.Join(", ", rule.EnumValues)}]", rule));
            }

            if (rule.Min is not null || rule.Max is not null)
            {
                if (TryGetNumeric(value, out var num))
                {
                    if (rule.Min is not null && num < rule.Min.Value)
                        violations.Add(new SchemaViolation(key,
                            $"Value {num} is below minimum {rule.Min.Value}", rule));
                    if (rule.Max is not null && num > rule.Max.Value)
                        violations.Add(new SchemaViolation(key,
                            $"Value {num} exceeds maximum {rule.Max.Value}", rule));
                }
            }
        }

        if (strict)
        {
            var schemaKeys = new HashSet<string>(schema.Keys);
            foreach (var flatKey in FlattenKeys(data))
            {
                if (!schemaKeys.Contains(flatKey))
                    violations.Add(new SchemaViolation(flatKey,
                        $"Unknown key '{flatKey}' not defined in schema",
                        new SchemaRule()));
            }
        }

        return violations;
    }

    /// <summary>
    /// Apply default values from schema to data (mutates data).
    /// </summary>
    public static void ApplyDefaults(
        Dictionary<string, object?> data,
        Dictionary<string, SchemaRule> schema)
    {
        foreach (var (key, rule) in schema)
        {
            if (rule.DefaultValue is not null && GetNested(data, key) is null)
            {
                SetNested(data, key, rule.DefaultValue);
            }
        }
    }

    /// <summary>
    /// Get the list of keys marked as sensitive in the schema.
    /// </summary>
    public static List<string> SensitiveKeys(Dictionary<string, SchemaRule> schema)
    {
        return schema
            .Where(kvp => kvp.Value.Sensitive)
            .Select(kvp => kvp.Key)
            .ToList();
    }

    /// <summary>
    /// Assert valid — throws <see cref="ValidationException"/> on failure.
    /// </summary>
    /// <exception cref="ValidationException">Thrown when validation fails.</exception>
    public static void AssertValid(
        Dictionary<string, object?> data,
        Dictionary<string, SchemaRule> schema,
        bool strict = false)
    {
        var violations = Validate(data, schema, strict);
        if (violations.Count > 0)
        {
            var messages = string.Join("; ", violations.Select(v => v.ToString()));
            throw new ValidationException(
                $"Schema validation failed: {messages}", violations);
        }
    }

    /// <summary>
    /// Get a nested value using dot-notation key.
    /// </summary>
    public static object? GetNested(Dictionary<string, object?> data, string key)
    {
        var parts = key.Split('.');
        object? current = data;

        foreach (var part in parts)
        {
            if (current is Dictionary<string, object?> dict && dict.TryGetValue(part, out var val))
                current = val;
            else
                return null;
        }

        return current;
    }

    /// <summary>
    /// Set a nested value using dot-notation key.
    /// </summary>
    public static void SetNested(Dictionary<string, object?> data, string key, object? value)
    {
        var parts = key.Split('.');
        var current = data;

        for (var i = 0; i < parts.Length - 1; i++)
        {
            if (!current.TryGetValue(parts[i], out var existing) ||
                existing is not Dictionary<string, object?> dict)
            {
                dict = new Dictionary<string, object?>();
                current[parts[i]] = dict;
            }
            current = dict;
        }

        current[parts[^1]] = value;
    }

    private static bool CheckType(object value, string type)
    {
        return type switch
        {
            "string" => value is string,
            "integer" => value is int or long or short or byte,
            "boolean" => value is bool,
            "number" => value is int or long or float or double or decimal or short or byte,
            "array" => value is System.Collections.IList,
            _ => true,
        };
    }

    private static bool CheckFormat(string value, string format)
    {
        if (FormatPatterns.TryGetValue(format, out var pattern))
            return Regex.IsMatch(value, pattern);

        return format switch
        {
            "url" => Uri.TryCreate(value, UriKind.Absolute, out var uri) &&
                     (uri.Scheme == "http" || uri.Scheme == "https"),
            "ip" => System.Net.IPAddress.TryParse(value, out _),
            "port" => int.TryParse(value, out var p) && p is >= 1 and <= 65535,
            _ => true,
        };
    }

    private static bool TryGetNumeric(object value, out double result)
    {
        result = 0;
        if (value is int i) { result = i; return true; }
        if (value is long l) { result = l; return true; }
        if (value is float f) { result = f; return true; }
        if (value is double d) { result = d; return true; }
        if (value is decimal dec) { result = (double)dec; return true; }
        return false;
    }

    private static List<string> FlattenKeys(Dictionary<string, object?> data, string prefix = "")
    {
        var result = new List<string>();
        foreach (var (key, value) in data)
        {
            var fullKey = string.IsNullOrEmpty(prefix) ? key : $"{prefix}.{key}";
            if (value is Dictionary<string, object?> nested)
                result.AddRange(FlattenKeys(nested, fullKey));
            else
                result.Add(fullKey);
        }
        return result;
    }
}
